File: code.js
const vertexShader = `
				struct Interpolators 
				{
					@builtin(position) position: vec4<f32>,
					@location(0) texcoord: vec2<f32>,
				};

				@vertex
				fn VSMain(@builtin(vertex_index) vertexId: u32) -> Interpolators
				{
					let vertices = array<vec2<f32>,3> (vec2(-1.0, -1.0), vec2(3.0, -1.0), vec2(-1.0, 3.0));
					return Interpolators(vec4(vertices[vertexId], 0.0, 1.0), vec2(0.5 * vertices[vertexId] + vec2(0.5, 0.5)));
				}
			`;
const fragmentShader = `
				struct Uniforms {iTime : f32};
				@group(0) @binding(0) var<uniform> uniforms : Uniforms;

				fn Triangle(p: vec2<f32>, a: vec2<f32>, b: vec2<f32>, c: vec2<f32>) -> f32
				{
					var ba = b - a;
					var cb = c - b;
					var ac = a - c;
					var pa = p - a;
					var pb = p - b;
					var pc = p - c;
					var q0: vec2<f32> = pa - ba * clamp( dot(pa,ba) / dot(ba,ba), 0.0, 1.0);
					var q1: vec2<f32> = pb - cb * clamp( dot(pb,cb) / dot(cb,cb), 0.0, 1.0);
					var q2: vec2<f32> = pc - ac * clamp( dot(pc,ac) / dot(ac,ac), 0.0, 1.0);   
					var s: f32 = ba.x * ac.y - ba.y * ac.x;
					var d = vec2<f32>(dot(q0, q0), s * (pa.x * ba.y - pa.y * ba.x));
					d = min(d, vec2<f32>(dot(q1, q1), s * (pb.x * cb.y - pb.y * cb.x)));
					d = min(d, vec2<f32>(dot(q2, q2), s * (pc.x * ac.y - pc.y * ac.x)));
					return step(-sqrt(d.x) * sign(d.y), 0.0);
				}

				@fragment
				fn PSMain(@location(0) texcoord: vec2<f32>) -> @location(0) vec4<f32> 
				{
					let t1 = Triangle(texcoord, vec2(0.050, 0.800), vec2(0.675, 0.800), vec2(0.350, 0.250));
					let t2 = Triangle(texcoord, vec2(0.675, 0.800), vec2(0.825, 0.525), vec2(0.512, 0.524));
					let t3 = Triangle(texcoord, vec2(0.513, 0.524), vec2(0.825, 0.525), vec2(0.675, 0.250));
					let t4 = Triangle(texcoord, vec2(0.746, 0.670), vec2(0.825, 0.800), vec2(0.910, 0.670));
					let t5 = Triangle(texcoord, vec2(0.746, 0.670), vec2(0.910, 0.670), vec2(0.825, 0.525));
					let c1 = vec3(0.00, 0.35, 0.61) * t1;
					let c2 = vec3(0.00, 0.40, 0.70) * t2;
					let c3 = vec3(0.00, 0.46, 0.80) * t3;
					let c4 = vec3(0.00, 0.57, 1.00) * t4;
					let c5 = vec3(0.00, 0.52, 0.91) * t5;
					let logo = c1 + c2 + c3 + c4 + c5;
					let f = logo.r > 0.0 || logo.g > 0.0 || logo.b > 0.0;
					return mix(vec4(logo, 1.0), vec4(vec3(sin(uniforms.iTime) * 0.5 + 0.5), 1.0), step(f32(f), 0.0));
				}
			`;
async function Main() {
  if (!navigator.gpu) {
    alert(
      "WebGPU is not supported, see https://webgpu.io or use Chrome Canary with chrome://flags/#enable-unsafe-webgpu",
    );
    return;
  }
  const device = await (await navigator.gpu.requestAdapter()).requestDevice();
  const format = navigator.gpu.getPreferredCanvasFormat();
  const renderPipeline = device.createRenderPipeline({
    layout: "auto",
    vertex: {
      module: device.createShaderModule({ code: vertexShader }),
      entryPoint: "VSMain",
    },
    fragment: {
      module: device.createShaderModule({ code: fragmentShader }),
      entryPoint: "PSMain",
      targets: [{ format }],
    },
    primitive: { topology: "triangle-list" },
  });
  const context = document.getElementById("canvas").getContext("webgpu");
  context.configure({ device, format, alphaMode: "premultiplied" });
  const uniformBuffer = device.createBuffer({
    size: 4,
    usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
  });
  const uniformBindGroup = device.createBindGroup({
    layout: renderPipeline.getBindGroupLayout(0),
    entries: [{ binding: 0, resource: { buffer: uniformBuffer } }],
  });
  const uniforms = new Float32Array(1);
  function Update(time) {
    uniforms[0] = time * 0.001;
    device.queue.writeBuffer(uniformBuffer, 0, uniforms);
    const commandEncoder = device.createCommandEncoder();
    const textureView = context.getCurrentTexture().createView();
    const renderPassColorAttachment = {
      view: textureView,
      clearValue: { r: 0, g: 0, b: 0, a: 1 },
      loadOp: "clear",
      storeOp: "store",
    };
    const renderPassEncoder = commandEncoder.beginRenderPass({
      colorAttachments: [renderPassColorAttachment],
    });
    renderPassEncoder.setPipeline(renderPipeline);
    renderPassEncoder.setBindGroup(0, uniformBindGroup);
    renderPassEncoder.draw(3, 1, 0, 0);
    renderPassEncoder.end();
    device.queue.submit([commandEncoder.finish()]);
    requestAnimationFrame(Update);
  }
  requestAnimationFrame(Update);
}
Main();

File: code.py
import js
from js import document, Float32Array, Int32Array
import pyodide.ffi
import sys


def to_js(value):
    return pyodide.ffi.to_js(value, dict_converter=js.Object.fromEntries)


def abort():
    js.alert("WebGPU is not supported")
    sys.exit(1)


def generate_data():
    if 1:
        from netgen.occ import unit_square

        m = unit_square.GenerateMesh(maxh=0.2)

        vertices = []
        for p in m.Points():
            for i in range(3):
                vertices.append(p[i])
            vertices.append(0)

        trigs = []
        edges = []
        for t in m.Elements2D():
            for i in range(3):
                trigs.append(t.vertices[i].nr - 1)
                edges.append(t.vertices[i].nr - 1)
                edges.append(t.vertices[(i + 1) % 3].nr - 1)
            trigs.append(0)
    else:
        vertices = [0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0]
        trigs = [0, 1, 2, 0]
        edges = [0, 1, 1, 2, 2, 0]

    vertex_array = Float32Array.new(vertices)
    trigs_array = Int32Array.new(trigs)
    edges_array = Int32Array.new(edges)

    return vertex_array, edges_array, trigs_array


async def main():

    if not js.navigator.gpu:
        abort()

    adapter = await js.navigator.gpu.requestAdapter()
    if not adapter:
        abort()

    device = await adapter.requestDevice()
    format = js.navigator.gpu.getPreferredCanvasFormat()

    # Get the canvas context and configure it
    canvas = document.getElementById("canvas")
    context = canvas.getContext("webgpu")
    context.configure(
        to_js(
            {
                "device": device,
                "format": format,
                "alphaMode": "premultiplied",
            }
        )
    )

    # Create the uniform buffer
    uniform_buffer = device.createBuffer(
        to_js(
            {
                "size": 4,
                "usage": js.GPUBufferUsage.UNIFORM | js.GPUBufferUsage.COPY_DST,
            }
        )
    )

    bindGroupLayout = device.createBindGroupLayout(
        to_js(
            {
                "entries": [
                    {
                        "binding": 0,
                        "visibility": js.GPUShaderStage.FRAGMENT,
                        "buffer": {"type": "uniform"},
                    },
                    {
                        "binding": 1,
                        "visibility": js.GPUShaderStage.VERTEX,
                        "buffer": {"type": "read-only-storage"},
                    },
                    {
                        "binding": 2,
                        "visibility": js.GPUShaderStage.VERTEX,
                        "buffer": {"type": "read-only-storage"},
                    },
                    {
                        "binding": 3,
                        "visibility": js.GPUShaderStage.VERTEX,
                        "buffer": {"type": "read-only-storage"},
                    },
                ],
            }
        )
    )

    vertices, edges, trigs = generate_data()

    vertex_buffer = device.createBuffer(
        to_js(
            {
                "size": vertices.length * 4,
                "usage": js.GPUBufferUsage.STORAGE | js.GPUBufferUsage.COPY_DST,
            }
        )
    )

    edges_buffer = device.createBuffer(
        to_js(
            {
                "size": edges.length * 4,
                "usage": js.GPUBufferUsage.STORAGE | js.GPUBufferUsage.COPY_DST,
            }
        )
    )

    trig_buffer = device.createBuffer(
        to_js(
            {
                "size": trigs.length * 4,
                "usage": js.GPUBufferUsage.STORAGE | js.GPUBufferUsage.COPY_DST,
            }
        )
    )

    device.queue.writeBuffer(vertex_buffer, 0, vertices)
    device.queue.writeBuffer(edges_buffer, 0, edges)
    device.queue.writeBuffer(trig_buffer, 0, trigs)

    # Create the bind group for the uniforms
    uniform_bind_group = device.createBindGroup(
        to_js(
            {
                "layout": bindGroupLayout,
                "entries": [
                    {"binding": 0, "resource": {"buffer": uniform_buffer}},
                    {"binding": 1, "resource": {"buffer": vertex_buffer}},
                    {"binding": 2, "resource": {"buffer": edges_buffer}},
                    {"binding": 3, "resource": {"buffer": trig_buffer}},
                ],
            }
        )
    )

    pipelineLayout = device.createPipelineLayout(
        to_js({"bindGroupLayouts": [bindGroupLayout]})
    )

    # Create the render pipeline
    shader_code = await (await js.fetch("./shader.wgsl")).text()
    render_pipeline_edges = device.createRenderPipeline(
        to_js(
            {
                "layout": pipelineLayout,
                "vertex": {
                    "module": device.createShaderModule(to_js({"code": shader_code})),
                    "entryPoint": "mainVertexEdge",
                },
                "fragment": {
                    "module": device.createShaderModule(to_js({"code": shader_code})),
                    "entryPoint": "mainFragmentEdge",
                    "targets": [{"format": format}],
                },
                "primitive": {"topology": "line-list"},
                "depthStencil": {
                    "format": "depth24plus",
                    "depthWriteEnabled": True,
                    "depthCompare": "less-equal",
                },
            }
        )
    )

    render_pipeline_trigs = device.createRenderPipeline(
        to_js(
            {
                "layout": pipelineLayout,
                "vertex": {
                    "module": device.createShaderModule(to_js({"code": shader_code})),
                    "entryPoint": "mainVertexTrig",
                },
                "fragment": {
                    "module": device.createShaderModule(to_js({"code": shader_code})),
                    "entryPoint": "mainFragmentTrig",
                    "targets": [{"format": format}],
                },
                "primitive": {"topology": "triangle-list"},
                "depthStencil": {
                    "format": "depth24plus",
                    "depthWriteEnabled": True,
                    "depthCompare": "less-equal",
                },
            }
        )
    )

    print("canvas size", canvas.width, canvas.height)
    depthTexture = device.createTexture(
        to_js(
            {
                "size": [canvas.width, canvas.height, 1],
                "format": "depth24plus",
                "usage": js.GPUTextureUsage.RENDER_ATTACHMENT,
            }
        )
    )

    uniforms = Float32Array.new(1)

    def update(time):
        uniforms[0] = time * 0.001
        device.queue.writeBuffer(uniform_buffer, 0, uniforms)

        command_encoder = device.createCommandEncoder()

        render_pass_encoder = command_encoder.beginRenderPass(
            to_js(
                {
                    "colorAttachments": [
                        {
                            "view": context.getCurrentTexture().createView(),
                            "clearValue": {"r": 1, "g": 1, "b": 1, "a": 1},
                            "loadOp": "clear",
                            "storeOp": "store",
                        }
                    ],
                    "depthStencilAttachment": {
                        "view": depthTexture.createView(to_js({"format": "depth24plus", "aspect": "all"})),
                        "depthLoadOp": "clear",
                        "depthStoreOp": "store",
                        "depthClearValue": 1.0,
                    },
                },
            )
        )
        render_pass_encoder.setViewport(0, 0, canvas.width, canvas.height, 0.0, 0.9999)

        render_pass_encoder.setPipeline(render_pipeline_edges)
        render_pass_encoder.setBindGroup(0, uniform_bind_group)
        render_pass_encoder.draw(edges.length, 1, 0, 0)
        render_pass_encoder.setPipeline(render_pipeline_trigs)
        render_pass_encoder.setBindGroup(0, uniform_bind_group)
        render_pass_encoder.draw(trigs.length, 1, 0, 0)
        render_pass_encoder.end()

        device.queue.submit([command_encoder.finish()])
        js.requestAnimationFrame(pyodide.ffi.create_proxy(update))

    js.requestAnimationFrame(pyodide.ffi.create_proxy(update))


# Run the main function
await main()
